R/staggregate.R

```diff
@@ -0,0 +1,601 @@
+# Function to convert raster to data.table from https://gist.github.com/etiennebr/9515738
+as.data.table.raster <- function(x, row.names = NULL, optional = FALSE, xy=FALSE, inmem = raster::canProcessInMemory(x, 2), ...) {
+  if(inmem) {
+    v <- data.table::as.data.table(raster::as.data.frame(x, row.names=row.names, optional=optional, xy=xy, ...))
+  } else {
+    tr <- raster::blockSize(x, n=2)
+    l <- lapply(1:tr$n, function(i)
+      data.table::as.data.table(raster::as.data.frame(raster::getValues(x,
+                                                                        row=tr$row[i],
+                                                                        nrows=tr$nrows[i]),
+                                                      row.names=row.names, optional=optional, xy=xy, ...)))
+    v <- data.table::rbindlist(l)
+  }
+  coln <- names(x)
+  if(xy) coln <- c("x", "y", coln)
+  data.table::setnames(v, coln)
+  v
+}
+
+# Function to convert raster to data.table and aggregate to daily values before transformation
+daily_aggregation <- function(data, daily_agg, overlay_weights){
+
+
+  # Data.table of weights
+  weights_dt <- data.table::as.data.table(overlay_weights)
+
+  # Extent of area weights with slight buffer to make sure all cells are included
+  min_x <- min(weights_dt$x) - 0.5
+  max_x <- max(weights_dt$x) + 0.5
+  min_y <- min(weights_dt$y) - 0.5
+  max_y <- max(weights_dt$y) + 0.5
+
+  weights_ext <- raster::extent(min_x, max_x, min_y, max_y)
+
+  ## Load climate data
+  ## -----------------------------------------------
+
+
+  # Immediately crop to weights extent
+  clim_raster <- raster::crop(raster::stack(data), weights_ext)
+
+  # Get layer names (dates)
+  all_layers <- names(clim_raster)
+  layer_names <- all_layers[seq(1, length(all_layers), 24)] # Keep every 24th layer name (1 per day)
+  layer_names <- paste0('year_', substring(layer_names, 2,5), 'month_', substring(layer_names, 7,8), '_day_', substring(layer_names, 10,11)) # Extract month/day for each layer
+
+  ## Aggregate to grid-day level
+  ## -----------------------------------------------
+
+  if(!daily_agg %in% c('average', 'sum')){
+
+    stop(crayon::red("Daily aggregation must be 'average' or 'sum'"))
+  }
+
+  ## Average
+  if(daily_agg == 'average'){
+
+    message(crayon::green("Averaging hourly values to get daily values"))
+
+    # Average over each set of 24 layers
+    indices<-rep(1:(raster::nlayers(clim_raster)/24),each=24)
+    clim_daily <- raster::stackApply(clim_raster, indices = indices, fun=mean) #Stack of 365/366 layers
+  }
+
+  ## Sum
+  if(daily_agg == 'sum'){
+
+    message(crayon::green("Summing hourly values to daily values"))
+
+    # Sum over each set of 24 layers
+    indices<-rep(1:(raster::nlayers(clim_raster)/24),each=24)
```

Comment:
Just flagging that these two lines (71-72) do not need to change. We are still indexing assuming that every set of 24 layers is one day. Not sure if this is the place Tamma was thinking would be useful to send a warning